reshape.py

```diff
@@ -184,72 +182,131 @@ def migrate_params(params):
 # 1. Find the bug in the `transpose` module. Unit-test it; fix; deploy.
 # 2. Copy/paste the `transpose` module's "render()" method here.
 # 3. Rename `render(...)` to `transpose(table, params)`
+# hard-code settings for now. TODO have Workbench pass render(..., settings=...)
+@dataclass
+class Settings:
+    MAX_COLUMNS_PER_TABLE: int
+    MAX_BYTES_PER_COLUMN_NAME: int
+
+
+settings = Settings(99, 120)
+
+
+@dataclass
+class GenColnamesResult:
+    names: List[str]
+    """All column names for the output table (even the first column)."""
+
+    warnings: List[str]
+    """All the things we should tell the user about how we tweaked names."""
+
+
+def _gen_colnames_and_warn(
+    first_colname: str, first_column: pd.Series
+) -> GenColnamesResult:
+    """
+    Generate transposed-table column names.
+
+    If `first_colname` is empty, `column.name` is the first output column. If
+    both are empty, auto-generate the column name (and warn).
+
+    Warn if ASCII-cleaning names, renaming duplicates, truncating names or
+    auto-generating names.
+
+    Assume `first_column` is text without nulls.
+    """
+    n_ascii_cleaned = 0
+    first_ascii_cleaned = None
+    n_default = 0
+    first_default = None
+    n_truncated = 0
+    first_truncated = None
+    n_numbered = 0
+    first_numbered = None
+
+    input_names = [first_colname or first_column.name]
+    input_names.extend(list(first_column.values))
+
+    names = []
+
+    for uccolname in gen_unique_clean_colnames(input_names, settings=settings):
+        name = uccolname.name
+        names.append(name)
+        if uccolname.is_ascii_cleaned:
+            if n_ascii_cleaned == 0:
+                first_ascii_cleaned = name
+            n_ascii_cleaned += 1
+        if uccolname.is_default:
+            if n_default == 0:
+                first_default = name
+            n_default += 1
+        if uccolname.is_truncated:
+            if n_truncated == 0:
+                first_truncated = name
+            n_truncated += 1
+        if uccolname.is_numbered:
+            if n_numbered == 0:
+                first_numbered = name
+            n_numbered += 1
+
+    warnings = []
+    if n_ascii_cleaned > 0:
+        warnings.append(
+            "Removed special characters from %d column names (see “%s”)"
```

Comment:
These warnings are going to be repeated in a bunch of modules. I think we should add some "default" i18n in the `cjwmodule.util.colnames` package. How about `gen_unique_clean_colnames_and_warn(names: List[str], *, existing_names: List[str] = [], settings: Settings = DefaultSetting()) -> Tuple[List[str], List[I18nMessage]]`? Its logic would be very similar to the logic here (except with five types of warning, not four).